src/interactions/panZoomInteraction.ts

```diff
@@ -20,56 +29,143 @@ export module Interaction {
-      if (this._xScale) {
-        this._zoom.x((<any> this._xScale)._d3Scale);
+    private _handleTouchMove(ids: number[], idToPoint: { [id: number]: Point; }, e: TouchEvent) {
+      if (this._touchIds.size() !== 2) {
+        return;
       }
-      if (this._yScale) {
-        this._zoom.y((<any> this._yScale)._d3Scale);
+      var points = this._touchIds.values();
+      var firstTouchPoint = points[0];
+      var secondTouchPoint = points[1];
+
+      var leftX = Math.min(firstTouchPoint.x, secondTouchPoint.x);
+      var rightX = Math.max(firstTouchPoint.x, secondTouchPoint.x);
+      var topY = Math.min(firstTouchPoint.y, secondTouchPoint.y);
+      var bottomY = Math.max(firstTouchPoint.y, secondTouchPoint.y);
+
+      var oldAvgX = (leftX + rightX) / 2;
+      var oldAvgY = (bottomY + topY) / 2;
+      var oldDiffX = rightX - leftX;
+      var oldDiffY = bottomY - topY;
+
+      ids.forEach((id) => {
+        var translatedP = this._translateToComponentSpace(idToPoint[id]);
+        if (this._isInsideComponent(translatedP)) {
+          this._touchIds.set(id.toString(), translatedP);
+        }
+      });
+
+      var newPoints = this._touchIds.values();
+      var newFirstTouchPoint = newPoints[0];
+      var newSecondTouchPoint = newPoints[1];
+
+      var newLeftX = Math.min(newFirstTouchPoint.x, newSecondTouchPoint.x);
+      var newRightX = Math.max(newFirstTouchPoint.x, newSecondTouchPoint.x);
+      var newTopY = Math.min(newFirstTouchPoint.y, newSecondTouchPoint.y);
+      var newBottomY = Math.max(newFirstTouchPoint.y, newSecondTouchPoint.y);
+
+      var newAvgX = (newLeftX + newRightX) / 2;
+      var newAvgY = (newBottomY + newTopY) / 2;
+      var newDiffX = newRightX - newLeftX;
+      var newDiffY = newBottomY - newTopY;
+
+      if (this._xScale != null && newDiffX !== 0 && oldDiffX !== 0) {
+        this._xScale.domain(PanZoom.magnify(this._xScale, oldDiffX / newDiffX, oldAvgX));
+        this._xScale.domain(PanZoom.translate(this._xScale, oldAvgX - newAvgX));
+      }
+      if (this._yScale != null && newDiffY !== 0 && oldDiffY !== 0) {
+        this._yScale.domain(PanZoom.magnify(this._yScale, oldDiffY / newDiffY, oldAvgY));
+        this._yScale.domain(PanZoom.translate(this._yScale, oldAvgY - newAvgY));
       }
-      this._zoom.on("zoom", () => this._rerenderZoomed());
-      this._zoom(this._hitBox);
     }
 
-    public _anchor(component: Component.AbstractComponent, hitBox: D3.Selection) {
-      super._anchor(component, hitBox);
-      this.resetZoom();
+    private _handleTouchEnd(ids: number[], idToPoint: { [id: number]: Point; }, e: TouchEvent) {
+      ids.forEach((id) => {
+        this._touchIds.remove(id.toString());
+      });
+      if (this._touchIds.size() < 2) {
+        this._setupDragInteraction();
+      }
     }
 
-    public _requiresHitbox() {
-      return true;
+    private static magnify<D>(scale: Scale.AbstractQuantitative<D>, magnifyAmount: number, centerValue: number) {
```

Comment:
Probably `magnifyScale` is a better name, since the method isn't on the `Scale`.